### PR TITLE
Issue 284 - Add custom classes for row and columns

### PR DESCRIPTION
**Pull request by rodfersou**

### CODE REVIEW

File: src/collective/cover/layout.py [hunk 1]
```diff
@@ -78,25 +78,28 @@ def grid_layout_edit(self, layout):
                     self.grid_layout_edit(element['children'])
 
     def render_section(self, section, mode):
-        if 'type' in section:
```
Comment (rodfersou) [inline src/collective/cover/layout.py]: Just refactor code, no changes in logic

File: src/collective/cover/layout_templates/group.pt [hunk 2]
```diff
@@ -5,36 +5,38 @@
       lang="en"
       i18n:domain="collective.cover"
       tal:omit-tag="">
-<body tal:omit-tag="">
```
Comment (rodfersou) [inline src/collective/cover/layout_templates/group.pt]: Reindent code

File: src/collective/cover/layout_templates/group.pt [hunk 3]
```diff
@@ -5,36 +5,38 @@
...
+    <body tal:omit-tag="">
+        <tal:mode define="mode options/mode">
+            <tal:block tal:define="section options/section | nothing"
+                       tal:condition="section">
+                <tal:layoutedit tal:condition="python:mode == 'layout_edit'">
+                    <div tal:define="klass python:section.get('class', '')"
+                         tal:attributes="class python:'%s %s' % (klass, '' if view.is_user_allowed_in_group() else 'disabled');
+                                         data-panel python: section.get('id', '');
+                                         data-column-size python: section.get('column-size', 1);
+                                         data-css-class python: section.get('css-class', None)">
```
Comment (rodfersou) [inline src/collective/cover/layout_templates/group.pt]: add data-css-class

File: src/collective/cover/layout_templates/group.pt [hunk 4]
```diff
@@ -5,36 +5,38 @@
...
-            <tal:layoutedit tal:condition="python:mode == 'layout_edit'">
-            <div    tal:define="klass python:section.get('class', '');
-                                data python:section.get('data', '')"
-                    tal:attributes="class python:'%s %s' % (klass, '' if view.is_user_allowed_in_group() else 'disabled');
-                                    data-panel python: section.get('id', '');
-                                    data-column-size python: data['column-size'] if 'column-size' in data else 1;
-                                    data-layout-type python: data['layout-type'] if 'layout-type' in data else 'column'">
+                        <a href="#" class="config-column-link">
+                            <i class="config-icon"></i>
+                        </a>
```
Comment (rodfersou) [inline src/collective/cover/layout_templates/group.pt]: add config icon

File: src/collective/cover/layout_templates/pagelayout.pt [hunk 5]
```diff
@@ -5,14 +5,11 @@
       lang="en"
       i18n:domain="collective.cover"
       tal:omit-tag="">
-<body tal:omit-tag="">
```
Comment (rodfersou) [inline src/collective/cover/layout_templates/pagelayout.pt]: reindent code

File: src/collective/cover/layout_templates/row.pt [hunk 6]
```diff
@@ -5,22 +5,39 @@
       lang="en"
       i18n:domain="collective.cover"
       tal:omit-tag="">
-<body tal:omit-tag="">
```
Comment (rodfersou) [inline src/collective/cover/layout_templates/row.pt]: reindent code

File: src/collective/cover/layout_templates/row.pt [hunk 7]
```diff
@@ -5,22 +5,39 @@
...
+                     tal:condition="section"
+                     tal:attributes="class python:'%s' % section.get('class', '');
+                                     data-css-class python: section.get('css-class', None)">
 
-            <tal:children repeat="child python:section.get('children') or []">
-                <div tal:replace="structure python:view.render_section(child, mode)" />
-            </tal:children>
+                    <a href="#" class="config-row-link">
+                        <i class="config-icon"></i>
+                    </a>
```
Comment (rodfersou) [inline src/collective/cover/layout_templates/row.pt]: add config icon

File: src/collective/cover/layout_templates/row.pt [hunk 8]
```diff
@@ -5,22 +5,39 @@
...
-            tal:condition="section"
-            tal:attributes="class python:'%s' % section.get('class', '');
-                            data-layout-type python: data['layout-type'] if 'layout-type' in data else 'row'">
+        <tal:mode define="mode options/mode;">
+            <tal:layoutedit tal:condition="python:mode == 'layout_edit'">
+                <div tal:define="section options/section | nothing;
+                                 data python:section.get('data', '') if section else None"
+                     tal:condition="section"
+                     tal:attributes="class python:'%s' % section.get('class', '');
+                                     data-css-class python: section.get('css-class', None)">
```
Comment (rodfersou) [inline src/collective/cover/layout_templates/row.pt]: add data-css-class

File: src/collective/cover/layout_templates/tile.pt [hunk 9]
```diff
@@ -5,73 +5,74 @@
       lang="en"
       i18n:domain="collective.cover"
       tal:omit-tag="">
-<body tal:omit-tag="">
```
Comment (rodfersou) [inline src/collective/cover/layout_templates/tile.pt]: reindent code

File: src/collective/cover/profiles/default/registry.xml [hunk 10]
```diff
@@ -10,10 +10,10 @@
...
...carousel", "type": "tile", "id": "d34acf7c70084016810946f9039ffb67"}]}]}, {"type": "row", "children": [{"type": "group", "data": {"column-size": 8, "layout-type": "column"}, "roles": ["Manager"], "children": [{"tile-type": "collective.cover.list", "type": "tile", "id": "7bb03c26b2504b8a9cadf5b852cc8a2c"}]}, {"type": "group", "data": {"column-size": 8, "layout-type": "column"}, "roles": ["Manager"], "children": [{"tile-type": "collective.cover.collection", "type": "tile", "id": "cc6d0ff688a143b38d64db2493ca74ce"}]}]}, {"type": "row", "children": [{"id": "group2", "type": "group", "data": {"column-size": 5, "layout-type": "column"}, "roles": ["Manager"], "children": [{"tile-type": "collective.cover.basic", "type": "tile", "id": "6673d5554e934b4396790a92698e32d0"}]}, {"type": "group", "data": {"column-size": 5, "layout-type": "column"}, "roles": ["Manager"], "children": [{"tile-type": "collective.cover.basic", "type": "tile", "id": "db093ff0b98c4475aeb4cb7c3a78e9ba"}]}, {"type": "group", "data": {"column-size": 5, "layout-type": "column"}, "roles": ["Manager"], "children": [{"tile-type": "collective.cover.basic", "type": "tile", "id": "bfe62e09db9e4a538f92dbeda8db42de"}]}]}]</element>
```
Comment (rodfersou) [inline src/collective/cover/profiles/default/registry.xml]: simplify layout

File: src/collective/cover/static/layout_edit.css [hunk 11]
```diff
@@ -258,6 +258,8 @@ input[type="submit"].btn::-moz-focus-inner {
     margin-left: 0 !important;
 }
 
+#content .layout .config-row-link:link,
```
Comment (rodfersou) [inline src/collective/cover/static/layout_edit.css]: add config icons

File: src/collective/cover/static/layout_edit.css [hunk 12]
```diff
@@ -273,6 +275,8 @@ input[type="submit"].btn::-moz-focus-inner {
     border: medium none;
 }
 
+#content .layout .config-row-link .config-icon,
```
Comment (rodfersou) [inline src/collective/cover/static/layout_edit.css]: add config icons

File: src/collective/cover/static/layout_edit.js [hunk 13]
```diff
@@ -1,27 +1,33 @@
 (function($) {
```
Comment (rodfersou) [inline src/collective/cover/static/layout_edit.js]: reindent code

File: src/collective/cover/static/layout_edit.js [hunk 14]
```diff
@@ -1,27 +1,33 @@
...
-                                  .attr('data-layout-type', 'tile'),
-            le = $('.layout'),
-            BeforeUnloadHandler;
+        n_columns = conf.ncolumns,
+        row_class = 'cover-row',
+        row_dom = $('<div class="'+ row_class +'">' +
+                    '    <a href="#" class="config-row-link">' +
+                    '        <i class="config-icon"></i>' +
+                    '    </a>' +
+                    '</div>'),
```
Comment (rodfersou) [inline src/collective/cover/static/layout_edit.js]: create row_dom in a more simple way to add config icon

File: src/collective/cover/static/layout_edit.js [hunk 15]
```diff
@@ -1,27 +1,33 @@
...
+        column_class = 'cover-column',
+        column_dom = $('<div class="'+ column_class +'">' +
+                       '    <a href="#" class="config-column-link">' +
+                       '        <i class="config-icon"></i>' +
+                       '    </a>' +
+                       '</div>'),
```
Comment (rodfersou) [inline src/collective/cover/static/layout_edit.js]: create column_dom in a more simple way to add config icon

File: src/collective/cover/static/layout_edit.js [hunk 16]
```diff
@@ -1,27 +1,33 @@
...
+        tile_class = 'cover-tile',
+        tile_dom = $('<div class="'+ tile_class +'">' +
+                     '</div>'),
```
Comment (rodfersou) [inline src/collective/cover/static/layout_edit.js]: follow same way to create tile_dom

File: src/collective/cover/static/layout_edit.js [hunk 17]
```diff
@@ -45,8 +51,6 @@
 
             setup: function() {
 
-                le.append('<div id="dialog" title="Resize Column"><p id="column-size-resize">Actual column size: <span></span></p><div id="slider"></div></div>');
```
Comment (rodfersou) [inline src/collective/cover/static/layout_edit.js]: I prefer to add this code with HTML, not JS

File: src/collective/cover/static/layout_edit.js [hunk 18]
```diff
@@ -84,6 +88,7 @@
                 self.generate_grid_css();
                 self.delete_manager();
                 self.resize_columns_manager();
+                self.class_chooser_manager();
```
Comment (rodfersou) [inline src/collective/cover/static/layout_edit.js]: new chooser css class chooser

File: src/collective/cover/static/layout_edit.js [hunk 19]
```diff
@@ -387,12 +392,12 @@
                 var resizer = $('<i/>').addClass('resizer');
                 $(columns).append(resizer);
 
-                $( "#dialog" ).dialog({
+                $( "#resizer" ).dialog({
```
Comment (rodfersou) [inline src/collective/cover/static/layout_edit.js]: better name for the dialog

File: src/collective/cover/static/layout_edit.js [hunk 20]
```diff
@@ -387,12 +392,12 @@
...
                     autoOpen: false
                 });
 
                 $( ".resizer" ).click(function() {
-                    $( "#dialog" ).dialog( "open" );
+                    $( "#resizer" ).dialog( "open" );
```
Comment (rodfersou) [inline src/collective/cover/static/layout_edit.js]: better name for the dialog

File: src/collective/cover/static/layout_edit.js [hunk 21]
```diff
@@ -512,16 +549,16 @@
...
                             entry.type = node_type[0];
                         }
+
                         if (node_type == 'column') {
                             entry.roles = ['Manager'];
                             entry.type = 'group';
-                            entry.data = {
-                                'column-size': $(this).data('columnSize'),
-                                'layout-type': $(this).data('layout-type')
+                            entry['column-size'] = $(this).data('columnSize');
```
Comment (rodfersou) [inline src/collective/cover/static/layout_edit.js]: layout-type was not needed and I took it out to simplify layout json

File: src/collective/cover/static/layout_edit.js [hunk 22]
```diff
@@ -512,16 +549,16 @@
...
+                        }
 
-                            };
+                        if ($(this).attr('data-css-class')) {
+                            entry['css-class'] = $(this).attr('data-css-class');
                         }
```
Comment (rodfersou) [inline src/collective/cover/static/layout_edit.js]: save css class

File: src/collective/cover/tests/test_cover.py [hunk 23]
```diff
@@ -141,7 +141,7 @@ def test_searchabletext_indexer(self):
                      "id": "test2",
```
Comment (rodfersou) [inline src/collective/cover/tests/test_cover.py]: simplify layout

File: src/collective/cover/tests/test_upgrades.py [hunk 24]
```diff
@@ -245,8 +245,8 @@ def test_new_uuids_structure(self):
         # simulate state on previous version
```
Comment (rodfersou) [inline src/collective/cover/tests/test_upgrades.py]: simplify layout

File: src/collective/cover/tests/test_upgrades.py [hunk 25]
```diff
@@ -364,3 +364,39 @@ def test_remove_orphan_annotations(self):
...
         self._do_upgrade_step(step)
         self.assertNotIn(foo, annotations)
+
+    def test_simplify_layout(self):
```
Comment (rodfersou) [inline src/collective/cover/tests/test_upgrades.py]: test upgrade step to move column-site to it's parent and remove data attribute (with layout-type in it)

File: src/collective/cover/widgets/selectpreview.py [hunk 26]
```diff
@@ -61,7 +61,7 @@ def simplify_layout(self, layout, simple_layout=[]):
```
Comment (rodfersou) [inline src/collective/cover/widgets/selectpreview.py]: simplify layout

File: src/collective/cover/browser/templates/layoutedit.pt [hunk 27]
```diff
@@ -5,82 +5,87 @@
...
+                        <li>
+                            <a id="btn-column" class="btn btn-small btn-inverse draggable"><i class="icon-white"></i><span i18n:translate="">Column</span></a>
+                        </li>
+                    </ul>
+                    <div tal:replace="structure context/@@tile_list"></div>
+                </div>
+                <div class="layout" tal:attributes="data-layoutmanager-settings view/layoutmanager_settings">
+                    <div tal:define="layout nocall:context/@@layout"
+                         tal:replace="structure layout/render_layout_edit" />
+                    <div id="resizer" title="Resize Column">
```
Comment (rodfersou) [inline src/collective/cover/browser/templates/layoutedit.pt]: Moved resizer div from JS to this template (as it must have one copy of it in the entire template)

File: src/collective/cover/browser/templates/layoutedit.pt [hunk 28]
```diff
@@ -5,82 +5,87 @@
...
+                        <p id="column-size-resize">
+                            Actual column size: <span></span>
+                        </p>
+                        <div id="slider"></div>
+                    </div>
+                    <div id="class-chooser" title="Choose Class">
```
Comment (rodfersou) [inline src/collective/cover/browser/templates/layoutedit.pt]: Dialog to select css class

**Comment by hvelarde** [comment]
@djowett do you want to try this and give us some feedback?

**Comment by djowett** [comment]
Trying to add a new cover gives me:

  Module zope.tales.tales, line 696, in evaluate
   - URL: /home/daniel/dev/Plone/Development/highfields_ph2.1/src/collective.cover/src/collective/cover/widgets/selectpreview_input.pt
   - Line 7, Column 0
   - Expression: <PathExpr standard:u'view/js'>
   - Names:
      {'args': (<SelectPreviewWidget 'form.widgets.template_layout'>,),
       'context': <PloneSite at /highfields_plone4>,
       'default': <object object at 0x7fba60053b70>,
       'loop': {},
       'nothing': None,
       'options': {},
       'repeat': {},
       'request': <HTTPRequest, URL=http://localhost:8082/highfields_plone4/++add++collective.cover.content>,
       'template': <zope.browserpage.viewpagetemplatefile.ViewPageTemplateFile object at 0x85c9650>,
       'view': <SelectPreviewWidget 'form.widgets.template_layout'>,
       'views': <zope.browserpage.viewpagetemplatefile.ViewMapper object at 0x7fba44ab9110>}
  Module zope.tales.expressions, line 217, in __call__
  Module zope.tales.expressions, line 211, in _eval
  Module collective.cover.widgets.selectpreview, line 45, in js
  Module collective.cover.widgets.selectpreview, line 73, in simplify_layout
  Module collective.cover.widgets.selectpreview, line 64, in simplify_layout
KeyError: 'column-size'
> /home/daniel/dev/Plone/Development/highfields_ph2.1/src/collective.cover/src/collective/cover/widgets/selectpreview.py(64)simplify_layout()
-> item['size'] = element['column-size']


I guess you want element['data']['column-size']

**Comment by rodfersou** [comment]
@djowett you will need to run upgrade step first :)

**Comment by djowett** [comment]
Ok.  I got it working - I needed to reinstall c.cover after running the upgrade step, but that may have been because I'd already run a prev version of the upgrade step...

Thoughts on this, as I said to Rodrigo in chat, I understand you don't want to give too much power to the user, so they can break the cover design with a badly thought out classname used elsewhere.  
*However*, I think using the vocabulary for tile classes looks confusing to the end user, and doesn't really solve your concern above.  

3 solutions (2 already suggested):

1. Use a textline in the popup for class and prefix the class
2. Use a textline in the popup for class and assign it to another attribute than class e.g. "data-row-name"
3. (My favourite), in "@@cover-settings" filter the styles as follows

Enter a list of styles to appear in the style pulldown. Format is title|className, one per line.
**Any className prefixed with 'tile-', 'row-' or 'column-' will only be available in the context of tiles, rows or columns respectively.**

**Comment by djowett** [comment]
One other note. Now we have 3 levels of config icons can we add a (hover) title to each link to say what each one does?

**Comment by hvelarde** [comment]
@djowett thanks for reviewing this; from your comments I deduce that the branch is working and I can merge it, right?

I think any other concerns/enhancements can be solved/implemented in other tickets before making the final release.

:+1: I like the idea of having a hover to the wheel to give some tips to the user.

I think is a good idea to add the possibility of filtering the classes based on a prefix because it leaves the door open to not including a prefix and using it for any object; @agnogueira comments?

**Comment by djowett** [comment]
MY custom grid system got broken... guess https://github.com/collective/collective.cover/blob/issue_284_b/docs/developer.rst#grid-systems needs updating.

**Comment by agnogueira** [comment]
-1 for the filter. If we need to separate the styles, we will need a extra config field for that. But I think we dont need separation. We are trying to fix problemns that can be solved in CSS. You can have different styles for the same classes when aplied to row, collumn or tile:

.row.alert { background: red;}
.cell.alert {background: lime;}
.tile.alert {background: blue;}

**Comment by djowett** [comment]
@agnogueira you're correct to say this can be solved in CSS, *but* that's not the point. The point is not to confuse the user.

The tile styles are used to *style* the tile - using styles like tile-dark-background etc
The row/column styles are, in from my point of view (and others too), used to give a handle for diazo etc that need to distinguish between rows. using styles like carouselrow, secondrow, thirds etc

You might have a different approach, but that's why I don't see them being the same, and when you present that to the end user it being confusing. Perhaps that explains the different perspectives?

+1 on hover for wheel before merge

**Comment by hvelarde** [comment]
@djowett open an issue for the custom grid system and paste your code so we can help you solving that.

can I merge the pull request now?

**Comment by djowett** [comment]
Grid sys is fixed. Do you want the patch for developer.rst before merge?

**Comment by agnogueira** [comment]
All final users here are looking to use styles into a row or column. They need the power to create boxes  (with border and bg) with a some tiles inside. Like the boxes here http://www.wired.com/ (business, design, entretainement...). 

If we use styles to do a different thing than style, we will confuse the end user.

I think you are focusing more in the developer side than the final user side. 
Maybe your need is  for a unique ID for row/collumn and not a class.

**Comment by hvelarde** [comment]
@djowett feel free to amend the documentation on master branch.

**Comment by djowett** [comment]
@agnogueira yes, I am focussed more on dev side (now you say it, I understand what @rodfersou meant by that), however these are the needs on my current project, that's why I put in the effort. Looking at #284 now, it is as you say focussed on styling, but #481 (where I started) is focussed on handles.  A unique ID for rows is what I'm after, but because CSS ids should be unique per doc, I thought class was better (or a different attribute as I mentioned earlier).  I suppose the mistake is as much mine as anyone's though & I think I can make it work for my case.

**Comment by hvelarde** [comment]
the problem could have been mixing the solution of 2 different use cases into one pull request.

[worse is better](https://en.wikipedia.org/wiki/Worse_is_better), in my opinion: #481 will add a lot of complexity to the package and I don't want to deal with that now, specially because that use case is easily solved with a workaround.

**Comment by djowett** [comment]
@hvelarde Yes true on 2 diff solutions. Not sure about worse is better - we might all end up using WP that way ;-)

See the tweak above to make it easier to add filters in another product (through overriding vocabulary) - will test override works another day though.

Thanks

**Comment by djowett** [comment]
@rodfersou @hvelarde can I just point out that users can't tell from the popup if they're editing Row or Column classes, so if they click inaccurately...

(other issue moving to #517)

**Comment by hvelarde** [comment]
@djowett worse is better: if you only need a blog then you must use WP, no matter that Plone is better for a full featured site.

please open a different PR for adding the context to the modal, that makes easier reviewing and merging.

thanks, again!

**Pull request merged**